File: 2do Curso/MP - Metodologia de la Programacion/Practica1/src/Fecha.cpp
#include "Fecha.h"


Fecha::Fecha( int dia, int mes, int anio)
{
    this->setFecha(dia,mes,anio);
    //ctor
}


void Fecha::setFecha( int dia, int mes, int anio)
{

    int dmax,diaMes[]= {0,31,28,31,30,31,30,31,31,30,31,30,31};
    this->anio=anio;
    if(this->bisiesto())
        diaMes[2]=29;
    if(mes<1)
        this->mes=1;
    else if(mes>12)
        this->mes=12;
    else
        this->mes=mes;
    dmax=diaMes[this->mes];

    if(dia>dmax)
        this->dia=dmax;
    else if(dia<1)
        this->dia=1;
    else
        this->dia=dia;
}

void Fecha::ver()const
{
    if(this->dia <10 )
        cout << "0";
    cout << this->dia << "/";
    if(this->mes <10)
        cout << "0";
    cout << this->mes << "/" << this->anio;


}


bool Fecha::bisiesto()const
{

    if(this->anio%400==0 || this->anio%4==0 && this->anio%100!=0)
        return true;
    else
        return false;

}

Fecha Fecha::operator++(int i)
{

    Fecha copia = *this; //guardo una copia de la fecha.
    //Incremento la fecha original.
    int dmax, diaMes[] = {0,31,28,31,30,31,30,31,31,30,31,30,31};
    if (this->bisiesto()) //si el año es bisiesto febrero tiene 29 dias
      diaMes[2]=29;
    dmax=diaMes[this->mes];
    this->dia++;
    if (this->dia>dmax) { //si al incrementar dia superamos el numero de dias de dicho mes
        this->dia=1;      //pasamos a 1
        this->mes++;      //del mes siguiente
            if (this->mes>12) { //si al incrementar mes pasamos de 12 meses
                this->mes=1;    //pasamos al mes 1
                this->anio++;   //del año siguiente
            }
    }
    return copia; // Devuelvo la copia antes de ser incrementada.

}

Fecha& Fecha::operator++()
{

    int dmax,diaMes[]= {0,31,28,31,30,31,30,31,31,30,31,30,31};
    if(this->bisiesto())
        diaMes[2]=29;
    dmax=diaMes[this->mes];
    this->dia++;
    if(this->dia > dmax)
    {
        this->dia=1;
        this->mes++;
        if(this->mes > 12)
        {
            this->mes=1;
            this->anio++;
        }
    }
    return *this;
}

Fecha Fecha::operator+(const int &x)const
{

    Fecha f=*this;
    for(int j=0;j<x;j++)
        f++;
    return f;
}

Fecha Fecha::operator+(const Fecha& f)const{

Fecha incr=*this;
incr.dia+=f.dia;
incr.mes+=f.mes;
incr.anio+=f.anio;

return incr;

}

Fecha operator+(int x,const Fecha& f)
{
    Fecha fech=f;
    for(int j=0;j<x;j++)
        ++fech;
    return fech;
}

ostream& operator<<(ostream& s,const Fecha& f){

 char *nombreMes[] = {"error", "ene", "feb", "mar", "abr", "may", "jun", "jul", "ago", "sep", "oct", "nov", "dic"};
  if (f.dia < 10)
    s << "0";
  s << f.dia << " ";

  s << nombreMes[f.mes] << " " << f.anio;


return s;

}
Fecha::~Fecha()
{
    //dtor
}

File: src/libClases/Fecha.java
package libClases;
import java.util.Scanner;

public final class Fecha implements Cloneable,Proceso {

	private int dia;
	private int mes;
	private int anio;
	
	public void setFecha(int d,int m,int a)
	{
		int dmax, diaMes[] = {31,28,31,30,31,30,31,31,30,31,30,31};
		 anio=a; //VIP debo asignar año para al llamar a bisiesto() tengo el año bien
		 if (m<1) //si el mes es incorrecto
		 m=1;
		 else if (m>12) //si el mes es incorrecto
		 m=12;
		 dmax=diaMes[m-1];
		 if (m==2 && bisiesto())
		 dmax++;
		 if (d>dmax)
		 d=dmax;
		 else if (d<1)
		 d=1;
		 dia=d;
		 mes=m;
	}
	
	public Fecha(int dia,int mes,int anio) 
	{
		setFecha(dia,mes,anio);
	}
	
	public Fecha(Fecha f)
	{
		dia=f.dia;
		mes=f.mes;
		anio=f.anio;
	}
	
	public int getDia() {return dia;};
	public int getMes() {return mes;};
	public int getAnio() {return anio;};
	public void setDia(int d) {dia=d;};
	public void setMes(int m) {mes=m;};
	public void setAnio(int a) {anio=a;};
	
	public String toString()
	{
		String s="";
		if(dia<10) s=s+0;
			s=s+dia+"/";
		if(mes<10) s=s+0;
		s=s+mes+"/"+anio;
		return s;
	}
	
	public boolean bisiesto()
	{
		boolean b=false;
		if(anio%4==0) 
			b=true;
		
		if(anio%100==0 && anio%400 !=0)
				b=false;
		
		return b;
	}
	
	public void ver()
	{
		System.out.println(this);
	}
	
	public static Fecha pedirFecha() {
		Fecha fecha=null;
		boolean valida=false;
		Scanner sc=new Scanner(System.in);
		int dia,mes,anio;
		
		do {
			System.out.println("Introduce la fecha (dd/mm/aaaa)");
			String cadena=sc.next();
			String[] tokens=cadena.split("/");
			try {
				if(tokens.length!=3)
					throw new NumberFormatException();
				dia=Integer.parseInt(tokens[0]);
				mes=Integer.parseInt(tokens[1]);
				anio=Integer.parseInt(tokens[2]);
				
				fecha= new Fecha(dia,mes,anio);
				
				if(fecha.getDia() != dia || fecha.getMes()!= mes)
					throw new NumberFormatException();
				
				valida=true;
				
			}catch(NumberFormatException e) {
				System.out.println("Fecha no valida");
			}
			
		}while(!valida);
		return fecha;
	}
	
	
	public static boolean mayor(Fecha f1,Fecha f2)
	{
		if(f1.anio*1000+f1.mes*100+f1.dia > f2.anio*1000+f2.mes*100+f2.dia)
			return true;
		else
			return false;
	}
	
	
	public Object Clone()
	{
		Object obj=null;
		try {
			obj=super.clone();
			
		}catch(CloneNotSupportedException ex) {
			System.out.println("No se puede duplicar");
		}
		return obj;
	}
	
	public boolean equals(Object obj)
	{
		if(this==obj)return true; //es el mismo objeto
		if(obj==null)return false;
		if(getClass()!=obj.getClass())
			return false; //si no son del mismo tipo de clase no son iguales
		Fecha c=(Fecha) obj; //casteo a fecha
		return (dia==c.dia && mes==c.mes && anio==c.anio);
	}
	
	public Fecha diaSig() {
		Fecha fechaSig = new Fecha(this.dia,this.mes,this.anio);
		
		int diaMes[] = {0,31,28,31,30,31,30,31,31,30,31,30,31};
	    if (bisiesto())                          //si el año es bisiesto febrero tiene 29 dias
	      diaMes[2]=29;

	    fechaSig.dia=this.dia+1;

	    if (fechaSig.dia>diaMes[fechaSig.mes]) {                          //si al incrementar dia superamos el numero de dias de dicho mes
	        do{
	        	fechaSig.dia=fechaSig.dia-diaMes[fechaSig.mes];             //pasamos al dia del mes anterior
	        	fechaSig.mes++;			                //incrementamos un mes
	            if (fechaSig.mes>12) {                    //si al incrementar mes pasamos de 12 meses
	            	fechaSig.mes=1;                       //pasamos al mes 1
	            	fechaSig.anio++;                      //del año siguiente
	                if (fechaSig.bisiesto())              //si el año es bisiesto febrero tiene 29 dias
	                    diaMes[2]=29;
	                else
	                    diaMes[2]=28;
	            }
	        }while(fechaSig.dia>diaMes[fechaSig.mes]);                    //hasta que el dia < diamax del mes en el q ns encontremos
	    }
		
		return fechaSig;
	}
	
	
	public static void main(String[] args) {
		 Fecha f1 = new Fecha(29,2,2001), f2 = new Fecha(f1), f3 = new Fecha(29,2,2004);
		 final Fecha f4=new Fecha(05,12,2003); //es constante la referencia f4
		 System.out.println("Fechas: " + f1.toString() + ", " + f2 + ", " + f3 + ", " + f4);
		 f1=new Fecha(31,12,2016); //31/12/2016
		 f4.setFecha(28, 2, 2008); //pero no es constante el objeto al que apunta
		 System.out.println(f1 +" "+ f2.toString() +" " + f3 + " "+ f4 + " "+ f1);
		 f1=new Fecha(f4.getDia()-10, f4.getMes(), f4.getAnio()-7); //f1=18/02/2001
		 f3=Fecha.pedirFecha(); // pide una fecha por teclado
		 if (f3.bisiesto() && Fecha.mayor(f2,f1))
		 System.out.println("El " + f3.getAnio() + " fue bisiesto, " + f1 + ", " + f3);
		 System.out.print("Fin\n");
		}
	
	
	}

